### ISSUE TITLE
How to get the fastest performance of the autodiff. 

### ISSUE THREAD

**Issue by taipinghu** [autodiff/autodiff#298]
Dear developers,

I am using the autodiff library to compute the gradient of a function that contains multiple loops.  I want to the calculate the gradient of the ewald_sum_real with respect to argument charges. In my case, the dimension of "charges" is about a few thousand, and the dimension of "natoms" is the same as that of "charges".

Now I have finished those calculations by calling autodiff (simply using gradient(ewald_sum_real, wrt, at)),  unfortunately, it is too slow.  I would like to know the best practices and tips to optimize the performance of the library in this scenario. 

Below is the code: 
```
void ewald_summation::ewald_sum_real(Eigen::VectorXd charges){
    double invsqrt2eta = 1.0 / (sqrt(2.0) * get_optimal_eta());
    double inv2eta = 1.0 / (2.0 * get_optimal_eta());
    double cutoff = get_optimal_cutoff_real();
    Eigen::Vector3i n = get_n_cells_t(box.cell, cutoff);
    int natoms = box.get_natoms();
    Eigen::Matrix3d cell = box.get_cell();
    Eigen::MatrixXd coord = box.get_atomic_coordinates();

    Eigen::VectorXd radius = box.mole.get_atomic_radius();
    Eigen::VectorXd hardness = box.mole.get_atomic_hardness();

    //cout << ewald.get_optimal_eta() << " " << cutoff << " " << n << endl;

    force_real.resize(box.get_natoms(), 3);
    force_real.setZero();
    energy_real = 0.0;

    int sj, sk, jats;
    Eigen::Vector3d dcell, d, ftmp;
    double d2, r, interf, tmp;

    for (int iat=0; iat<natoms; iat++) {
        for (int i=0; i<=n(0); i++) {
            sj = -n(1);
            if (i==0) sj = 0;
            for (int j=sj; j<=n(1); j++){
                sk = -n(2);
                if (i==0 && j==0) sk = 0;
                for (int k=sk; k<=n(2); k++){
                    dcell = i*cell.row(0) + j*cell.row(1) + k*cell.row(2);
                    jats = 0;
                    if (i==0 && j==0 && k==0) jats = iat+1;
                    for (int jat=jats; jat<natoms; jat++){
                        if (i!=0 || j!=0 || k!=0 || iat!=jat) {
                            d(0) = coord(iat, 0) - coord(jat, 0) + dcell(0);
                            d(1) = coord(iat, 1) - coord(jat, 1) + dcell(1);
                            d(2) = coord(iat, 2) - coord(jat, 2) + dcell(2);
                            d2 = d.squaredNorm();
                            if (d2 <= cutoff*cutoff) {
                                r      = sqrt(d2);
                                interf = erfc(r * invsqrt2eta);
                                tmp    = sqrt(pow(radius(iat), 2)+pow(radius(jat), 2));
                                interf = interf - erfc(r/(sqrt(2.0)*tmp));
                                energy_real += 2.0 * charges(iat) * charges(jat) * interf / r;
                            }
                        }
                    }
                }
            }
        }
        energy_real += charges(iat)*charges(iat)/(radius(iat)*2.0*sqrt(PI)) * 2.0;
    }
    energy_real = energy_real / 2.0;
}
```

Are there any specific techniques or optimizations that I should consider when using the autodiff library for such cases? 

I would appreciate any advice or examples you could provide.

best regards,
taiping

**Comment by allanleal** [comment]
Hi @taipinghu , what is the number type you have tried? Also, from your code above, it does not seem difficult to compute `energy_real` as well as its derivatives with respect to `charge[i]`: 

~~~
energy_real += 2.0 * charges(iat) * charges(jat) * interf / r;
energy_real_charges[iat] += 2.0 * charges(jat) * interf / r;
energy_real_charges[jat] += 2.0 * charges(iat) * interf / r;
~~~

Please double check the derivatives above!

Using automatic differentiation with forward number types `dual` and `real` will require the reevaluation of your sum method once for each variable in `charges`. If only a few variables in `charges` matter, you could then compute derivatives with respect to them only (instead of thousands).

Using reverse mode, `var` type, you could compute the derivatives of your energy variable with respect to all charges in a single call. However, `var` creates an expression tree (in runtime) that is slower to evaluate than `dual` and `real`.

**Comment by taipinghu** [comment]
Dear @allanleal ,

thanks for your quick reply.  

I'm very sorry for your misunderstanding caused by my unclear description. Let me rephrase my question:
(1) I have a complex energy summation form with multiple loops (the so-called Ewald summation for the Coulomb Interaction), and its expression is: 
![image](https://github.com/autodiff/autodiff/assets/32788428/f409c0f0-c44d-4351-9782-bb8390e79beb)
(2)Charges are unknown and need to be solved by me, and they satisfy the known sum of all charges, Q;
(3)I will use the Lagrange multiplier method to minimize the Lagrangian in order to obtain the charges.
Therefore, the problem becomes finding the root of the derivative of the Lagrangian. 

In fact, we also can take the seceond-order derivative of the energy with respect to the charges and explicitly construt the Hessian matrix, but this is very expensive. Therefore, I think that using automatic differentiation techniques to find the gradient of the Lagrangian will be very convenient. 

In fact, I have finished above process in JAX framework, which is a Python code. However, even in the GPU, the effeciency is unacceptable.  

I want to implement it in the C++ framework and I will further interface those code to an another big program (writen in C++). 

I'm not sure if my description is clear, but if necessary, I can send you my code and the problem privately. 

Sincerely thanks a lot.

regards, 
taiping
